src/Type/Equality.purs

```diff
@@ -14,10 +16,26 @@ module Type.Equality
 -- | Note: any instance will necessarily overlap with
 -- | `refl` below, so instances of this class should
 -- | not be defined in libraries.
+class TypeEquals :: forall k. k -> k -> Constraint
 class TypeEquals a b | a -> b, b -> a where
-  to :: a -> b
-  from :: b -> a
+  leibniz :: forall p. p a -> p b
 
 instance refl :: TypeEquals a a where
-  to a = a
-  from a = a
+  leibniz a = a
+
+newtype Op :: forall k. k -> k -> Type
+newtype Op a b = Op (forall p. p b -> p a)
+
+leibnizOp :: forall p a b. TypeEquals a b => p b -> p a
+leibnizOp = case leibniz (Op (\pa -> pa) :: Op a a) of
+  (Op f :: Op a b) -> f
+
+newtype To a = To a
+
+to :: forall a b. TypeEquals a b => a -> b
+to a = case leibniz (To a) of
+  To b -> b
```

Comment:
The `To` newtype is perhaps better for readability but we could define `to`  as `leibniz \a -> a`.

Reply:
Good point, I missed that that was possible.

Reply:
Good point, I missed that that was possible.